File: task2.go
package main

import (
	"fmt"
)

func isMatch(s string, p string, sI int, pI int, res [][]int) bool {
	if pI == len(p) {
		return sI == len(s)
	}

	if sI == len(s) {
		return p[pI] == '*' && isMatch(s, p, sI, pI+1, res)
	}

	if res[sI][pI] != -1 {
		return res[sI][pI] == 1
	}

	var match bool
	if p[pI] == '*' {
		match = isMatch(s, p, sI+1, pI, res) || isMatch(s, p, sI, pI+1, res)
	} else {
		match = (s[sI] == p[pI] || p[pI] == '?') && isMatch(s, p, sI+1, pI+1, res)
	}

	res[sI][pI] = 0
	if match {
		res[sI][pI] = 1
	}

	return match
}

func isPatternMatch(s string, p string) bool {
	res := make([][]int, len(s)+1)
	for i := range res {
		res[i] = make([]int, len(p)+1)
		for j := range res[i] {
			res[i][j] = -1
		}
	}
	return isMatch(s, p, 0, 0, res)
}

func main() {
	var s, p string
	fmt.Println("Введите s:")
	fmt.Scan(&s)
	fmt.Println("Введите p:")
	fmt.Scan(&p)

	if isPatternMatch(s, p) {
		fmt.Println("true")
	} else {
		fmt.Println("false")
	}
}

File: task2.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

bool isMatch(const string& s, const string& p, int sI, int pI, vector<vector<int>>& res) {
    if (pI == p.length())
        return sI == s.length();

    // Если мы достигли конца строки s, но остались символы в p, проверяем, являются ли они всеми '*'
    if (sI == s.length())
        return p[pI] == '*' && isMatch(s, p, sI, pI + 1, res);

    // Если мы ранее проверяли эту комбинацию индексов, возвращаем результат из памяти
    if (res[sI][pI] != -1)
        return res[sI][pI];

    bool match;
    // Обрабатываем случай с *
    if (p[pI] == '*') {
        // '*' соответствует пустой последовательности или последовательности из одного или нескольких символов
        match = isMatch(s, p, sI + 1, pI, res) || isMatch(s, p, sI, pI + 1, res);
    }
    else {
        // Обычное сравнение символов или сопоставление с подстановочным знаком '?'
        match = (s[sI] == p[pI] || p[pI] == '?') && isMatch(s, p, sI + 1, pI + 1, res);
    }

    // Запоминаем результат для данной комбинации индексов
    return res[sI][pI] = match;
}


bool isPatternMatch(const string& s, const string& p) {
    vector<vector<int>> res(s.length() + 1, vector<int>(p.length() + 1, -1)); // Матрица для хранения значений
    return isMatch(s, p, 0, 0, res);
}

int main() {
    system("chcp 65001");
    string s;
    string p;
    cout << "Введите s: ";
    cin  >> s;
    cout << endl;
    cout << "Введите p: ";
    cin >> p;
    cout << endl;
    if (isPatternMatch(s, p))
        cout << "true";
    else
        cout << "false";

    return 0;
}